### PR TITLE
Vlad_Protosevich_protosevich2001@gmail.com

### PR DESCRIPTION
**Pull request by protosevichVlad**
Iteration 3
Add README and bug fixes

### CODE REVIEW

File: final_task/README.md [hunk 1]
```diff
@@ -1,3 +1,38 @@
...
+  source         RSS URL
+
+optional arguments:
+  -h, --help     show this help message and exit
+  --version      Print version info
+  --json         Print result as JSON in stdout
+  --verbose      Outputs verbose status messages
+  --limit LIMIT  Limit news topics if this parameter provided
+  --date DATE    Obtaining the cached news without the Internet
+```
```
Comment (HenadziStantchik) [inline final_task/README.md:159]: It would be nice to see a bit more details about your implementation here specifically.

File: final_task/rss_reader/rss_reader.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    'feed': [
+        'title'
+    ],
+    'entries': [
+        'title',
+        'published',
+        'link',
+        'summary'
+    ]
+}
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: In genereal it is better to avoid global variables whenever you can. You could just pass all of them as arguments to the corresponfing functions instead

File: final_task/rss_reader/rss_reader.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            return f'HTTP Status Code {data.status}'
+    except Exception as exc:
+        return f'ERROR: {exc}'
+
+
+def to_json(data) -> json:
+    """convert data to JSON format"""
+    result = {}
+    if isinstance(data, feedparser.FeedParserDict):
+        if 'feed' in PRINT_SETTING.keys():
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: A question:

Why do you need this check if you yourself created this dict with this key at the top of the module?
Same question for check a bit lower.

File: final_task/rss_reader/rss_reader.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    global LINKS
+    image = ''
+    if '<' not in string:
+        return html.unescape(string)
+    if 'img' in string:
+        image_link_and_alt_text = get_img(string)
+        LINKS.append(image_link_and_alt_text[0])
+        image = '[image ' + str(len(LINKS)) + ': ' + image_link_and_alt_text[1] + '][' + str(len(LINKS)) + '] '
+    string = string[string.find('<') + 1:]
+    while string.find('<') - string.find('>') < 2:
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: This line can under some circumstances freeze your program. See point 6 of my feedback.

**Comment by HenadziStantchik** [comment]
Here is feedback about your script work:
1. You cannot get the app version without specifying source RSS (e.g. `python rss_reader.py --version` does not work)
2. You have encoding problem:

image

3. While you are parsing news correctly, it would be better if you at least numerate links to the images. You got them numerated in news feed, but in links they are not. Or for the future implementations it would be better group them like this:
```
Title: ...
Date: ...
Link: ...
Description: ...
Links:
[1] - image_1_url
[2] - image_2_url
...
```

This way you could awoid potential mistakes in the future convertations to other formats.

4. `--verbose` option does not work.

5. If `--limit` was not specified by user then you should print all available feed.

6. Your script freezes if `--limit` equals a certain number.
(A hint: try different large `--limit` and see what happens) 

7. `--limit` does not work on number of news when using `--date`

8. If I use reader with `--limit 5` and then with `--limit 1` your script will save only 1 piece of news to the cache.

upd: please be sure to use labels in the future. Check slack for more info.

**Comment by protosevichVlad** [comment]
Iteration 3

**Comment by HenadziStantchik** [comment]
I am getting `ModuleNotFoundError` by trying to launch your app without using distribution tools. Please fix it.

File: final_task/rss_reader/rss_reader.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument("source", help="RSS URL", nargs='?', default='', type=str)
+    parser.add_argument("--version", help="Print version info", action="store_true")
+    parser.add_argument("--json", help="Print result as JSON in stdout", action="store_true")
+    parser.add_argument("--verbose", help="Outputs verbose status messages", action="store_true")
+    parser.add_argument("--limit", help="Limit news topics if this parameter provided", type=int)
+    parser.add_argument("--date", help="Obtaining the cached news without the Internet", type=str)
+    args = parser.parse_args()
+    if not args.limit:
+        args.limit = 1000
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: And what if there will be 1001 entry in the feed? If user did not specify the limit it is better to link it somehow with feed size

File: final_task/rss_reader/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+__version__ = '3.6'
```
Comment (HenadziStantchik) [inline final_task/rss_reader/__init__.py]: Usuall it is better to declare global constants using uppercase, e.g. `VERSION`

File: final_task/rss_reader/rss_reader.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        data = read_feed_form_file(args.date)
+        if args.json:
+            data = limited_json(data, args.limit)
+            print(json.dumps(data, ensure_ascii=False))
+        else:
+            print(get_string_with_result(data, args.limit))
+    elif args.source:
+        data = get_object_feed(args.source)
+        data = to_json(data)
+        if 'error' not in data:
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: It is better to raise an exception if corresponding function and handle it here, instead of checking if your data is normal data or error message.

File: final_task/rss_reader/work_with_text.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                if isinstance(item, dict):
+                    for key_l, value_l in item.items():
+                        result += edit_key(key_l) + value_l
+                        result += '\n'
+                else:
+                    result += f'[{index + 1}] - {item}'
+                result += '\n'
+    else:
+        result = data
+    return result
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_text.py]: If I understood the purpose of this function correctly: `json.dumps` has an keyword argument `indent`. It is better to use this instead of this function.

File: final_task/rss_reader/work_with_text.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+def edit_key(input_key: str) -> str:
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_text.py:49]: A docstring would be good here or at least please make function name more informative

**Comment by HenadziStantchik** [comment]
> Here is feedback about your script work:
> 
> 1. You cannot get the app version without specifying source RSS (e.g. `python rss_reader.py --version` does not work)
> 2. You have encoding problem:
> 
> image
> 
> 1. While you are parsing news correctly, it would be better if you at least numerate links to the images. You got them numerated in news feed, but in links they are not. Or for the future implementations it would be better group them like this:
> 
> ```
> Title: ...
> Date: ...
> Link: ...
> Description: ...
> Links:
> [1] - image_1_url
> [2] - image_2_url
> ...
> ```
> 
> This way you could awoid potential mistakes in the future convertations to other formats.
> 
> 1. `--verbose` option does not work.
> 2. If `--limit` was not specified by user then you should print all available feed.
> 3. Your script freezes if `--limit` equals a certain number.
>    (A hint: try different large `--limit` and see what happens)
> 4. `--limit` does not work on number of news when using `--date`
> 5. If I use reader with `--limit 5` and then with `--limit 1` your script will save only 1 piece of news to the cache.
> 
> upd: please be sure to use labels in the future. Check slack for more info.

Also please respond in comments what of these points did you fix.

**Comment by protosevichVlad** [comment]
> I am getting `ModuleNotFoundError` by trying to launch your app without using distribution tools. Please fix it.

I run my app using `run_pakege.py`, another way I have not yet found. If I do run through `rss_reader.py` the I have not works after installations through `setup.py`

**Comment by protosevichVlad** [comment]
> > Here is feedback about your script work:
> > 
> > 1. You cannot get the app version without specifying source RSS (e.g. `python rss_reader.py --version` does not work)
> > 2. You have encoding problem:
> > 
> > image
> > 
> > 1. While you are parsing news correctly, it would be better if you at least numerate links to the images. You got them numerated in news feed, but in links they are not. Or for the future implementations it would be better group them like this:
> > 
> > ```
> > Title: ...
> > Date: ...
> > Link: ...
> > Description: ...
> > Links:
> > [1] - image_1_url
> > [2] - image_2_url
> > ...
> > ```
> > 
> > 
> > This way you could awoid potential mistakes in the future convertations to other formats.
> > 
> > 1. `--verbose` option does not work.
> > 2. If `--limit` was not specified by user then you should print all available feed.
> > 3. Your script freezes if `--limit` equals a certain number.
> >    (A hint: try different large `--limit` and see what happens)
> > 4. `--limit` does not work on number of news when using `--date`
> > 5. If I use reader with `--limit 5` and then with `--limit 1` your script will save only 1 piece of news to the cache.
> > 
> > upd: please be sure to use labels in the future. Check slack for more info.
> 
> Also please respond in comments what of these points did you fix.

--version - fixed
encoding problem - fixed
--verbose - fixed
--limit - fixed

**Comment by HenadziStantchik** [comment]
Here is some feedback on your application:

1 (optional). It would be great to add some more logging to printong functions.

2. `--date` argument stopped working completely. instead I get `FileNotFoundError` with traceback to stdout.

Please address existing issues

File: final_task/rss_reader/requirements.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+feedparser==5.2.1
```
Comment (HenadziStantchik) [inline final_task/rss_reader/requirements.txt]: Please include 3rd-party libraries you are using here

File: final_task/README.md [hunk 11]
```diff
@@ -1,3 +1,79 @@
...
+}
+```
+
+* "title" - source name\
+* "items" - list with dictionary which contains one news information\
+    * "title" - headline news
+    * "published" - date publication
+    * "link" - link
+    * "summary" - description
+* "links" - this is a list with links to the image of the I-th news
```
Comment (HenadziStantchik) [inline final_task/README.md]: Please specify the way you store news entries here.

**Comment by HenadziStantchik** [comment]
Here is some feedback on your app work:

1. Date specified by user in `--date` option should be publication date, not the date you parsed news and stored them

2. HTML-convertation has encoding problems:

protosevich

3. Convertation to another format from local cache does not work without internet connection. Moreover it throws exception `ConnectionError` with traceback to stdout.
It should be this way: if you convert news from local storage and have no internet connection to download images, just instert links to the images instead.

4. If you will fetch rss feed from another source, news entries from the old source in local cache will disappear. News in the local cache should be stored indefinitely.

5. (optional) You can make so that if RSS url specified with `--date`, then your app will look news in the local cache from this specific rss feed.

File: final_task/rss_reader/decorators.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import logging
+
+
+def functions_log(f):
```
Comment (HenadziStantchik) [inline final_task/rss_reader/decorators.py]: Please avoid one-letter variable names

File: final_task/rss_reader/rss_reader.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    elif args.date:
+        data = work_with_file.read_feed_form_file(args.date)
+    elif args.source:
+        data = get_object_feed(args.source)
+        data = work_with_dict.to_dict(data)
+        if 'error' not in data:
+            work_with_file.add_feed_to_file(data)
+    else:
+        print('How work with application?\nEnter in command line: rss-reader -h')
+        exit(1)
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: It is better to have only one exit point from the application. You just can raise your custom exception here and handle it lower

File: final_task/rss_reader/work_with_colorize.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        print(('{0} ' + data['error'] + ' {1}').format(red, default))
+    print(('{0} ' + data['title'] + '\n').format(yellow))
+    for index_news, dict_news in enumerate(data['items']):
+        print(('{0}Title: {1}' + dict_news['title']).format(green, red))
+        print(('{0}Date: {1}' + dict_news['published']).format(green, pink))
+        print(('{0}Link: {1}' + dict_news['link']).format(green, blue))
+        print(('{0}Description: {1}' + dict_news['summary'] + '{2}\n').format(green, description_color, default))
+    print('{0}Links:'.format(colored.fg(89)))
+    for index_links, link in enumerate(data['links']):
+        print(('{0}[' + str(index_links+1) + '] - {1}' + link + '{2}').format(green, blue, default))
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_colorize.py]: It is better to use `formatted string` for this. It will make code a bit more readable.

File: final_task/rss_reader/work_with_html.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import os
+import datetime
+
+
+def write_to_html_file(data: dict, path: str):
+    if os.path.isdir(path):
+        date_str = datetime.datetime.now().date().strftime('%Y%m%d')
+        filename = path + os.path.sep + date_str + '_' + data['title'][:data['title'].find(' ')].replace(':', '') + '.html'
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_html.py]: it is better to just use `os.path.join` instead of manually concatenate multiple strings into one using `os.path.sep`

File: final_task/rss_reader/work_with_pdf.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        effective_page_width = pdf.w - 2 * pdf.l_margin
+        pdf.compress = False
+        pdf.add_page()
+        pdf.add_font("my", '', '18223.ttf', uni=True)
+        pdf.set_font("my", size=30)
+        pdf.cell(w=0, txt=data['title'])
+        pdf.set_font("my", size=20)
+        pdf.ln(30)
+        pdf.set_line_width(1)
+        pdf.set_draw_color(255, 0, 0)
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_pdf.py]: It is more logical to move this code to a separate function.

File: final_task/rss_reader/work_with_pdf.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        for index_news, news_dict in enumerate(data['items']):
+            pdf.line(20, pdf.get_y() - 10, effective_page_width, pdf.get_y() - 10)
+            pdf.multi_cell(effective_page_width, 10, news_dict['title'])
+            if news_dict['summary'][0] == '[' and data['links'][index_news]:
+                p = requests.get(data['links'][index_news])
+                if p.status_code == 200 and data['links'][index_news].find('.gif') == -1:
+                    out = open('img' + str(index_news) + '.jpg', "wb")
+                    out.write(p.content)
+                    out.close()
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_pdf.py]: It is better to use context manager for this

File: final_task/rss_reader/work_with_text.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    #         for index, item in (enumerate(value) if len(value) < limit else enumerate(value[:limit])):
+    #             if isinstance(item, dict):
+    #                 for key_l, value_l in item.items():
+    #                     result += edit_key(key_l) + value_l
+    #                     result += '\n'
+    #             else:
+    #                 result += f'[{index + 1}] - {item}'
+    #             result += '\n'
+    # else:
+    #     result = data
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_text.py]: Please avoid to leave commented-out code in a production code.

File: final_task/rss_reader/MyException.py [hunk 19]
```diff
@@ -0,0 +1,3 @@
+class MyException(Exception):
```
Comment (HenadziStantchik) [inline final_task/rss_reader/MyException.py]: Please avoid using names such as `my...` in the production code. This exception class name should briefly describe the situation when you want to raise it.

> Please avoid using names such as `my...` in the production code.

The same for module names

File: final_task/rss_reader/rss_reader.py [hunk 20]
```diff
@@ -25,15 +26,15 @@ def get_object_feed(url: str) -> Union[str, feedparser.FeedParserDict]:
...
                 return data
         else:
-            return f'HTTP Status Code {data.status}'
+            raise MyException.MyException(f'HTTP Status Code {data.status}')
     except AttributeError:
-        return f'ERROR: {url} - is not url(example url "https://google.com")'
+        raise MyException.MyException(f'{url} - is not url(example url "https://google.com")')
     except Exception as exc:
-        return f'ERROR: {exc}'
+        raise MyException.MyException(exc)
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py]: It is better to have multiple custom exceptions for different situations.

File: final_task/rss_reader/work_with_file.py [hunk 21]
```diff
@@ -2,40 +2,65 @@
...
 @decorators.functions_log
 def read_feed_form_file(date_str: str):
     if os.path.exists('cache.txt'):
-        with open('cache.txt', 'r') as file:
-            for line in file:
-                if date_str in line:
-                    return json.loads(line[line.find(date_str) + len(date_str) + 1:].encode('UTF-8'))
-            else:
-                return 'Date ' + date_str + ' not found in cache.'
+        with open('cache.txt', 'r') as read_file:
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_file.py]: If you store news in JSON format anyway, why not name the file `cache.json`?

File: final_task/rss_reader/work_with_file.py [hunk 22]
```diff
@@ -2,40 +2,65 @@
...
+        'Apr': '04',
+        'May': '05',
+        'June': '06',
+        'July': '07',
+        'Aug': '08',
+        'Sept': '09',
+        'Oct': '10',
+        'Nov': '11',
+        'Dec': '12',
+    }
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_file.py]: It is better to use `datetime` standard library for this

File: final_task/rss_reader/work_with_html.py [hunk 23]
```diff
@@ -1,16 +1,19 @@
...
 import datetime
+import MyException
 
 
 def write_to_html_file(data: dict, path: str):
     if os.path.isdir(path):
         date_str = datetime.datetime.now().date().strftime('%Y%m%d')
-        filename = path + os.path.sep + date_str + '_' + data['title'][:data['title'].find(' ')].replace(':', '') + '.html'
+        filename = os.path.join(path, date_str + '_' +
+                                data['title'][:data['title'].find(' ')].replace(':', '').replace('.', '') + '.html')
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_html.py:10]: It is better to split this line into two:
```
date_str = date_str + '_' ....
filename = os.path.join(....
```

**Comment by HenadziStantchik** [comment]
> 2\. HTML-convertation has encoding problems:
> 
> protosevich

Sometimes there are empty lines in the image url list:

protosevich

Please ensure that your app works with RSS feed urls other than yahoo.
For any other RSS url I tried I got an `ERROR: HTTP STATUS CODE 301`

Also please describe the issues you fixed either in the commit message or in the comments here

**Comment by protosevichVlad** [comment]
> Sometimes there are empty lines in the image url list:
> 
> protosevich
> 
> Please ensure that your app works with RSS feed urls other than yahoo.
> For any other RSS url I tried I got an `ERROR: HTTP STATUS CODE 301`

Fixed

**Comment by protosevichVlad** [comment]
> 2. HTML-convertation has encoding problems:
> 
> protosevich

If you will see symbols, take screenshots these symbols in cmd(terminal) or site

**Comment by HenadziStantchik** [comment]
> > 1. HTML-convertation has encoding problems:
> > 
> > protosevich
> 
> If you will see symbols, take screenshots these symbols in cmd(terminal) or site

Problem still persists

**Comment by protosevichVlad** [comment]
The problem is the symbol `'`. After checking in different browsers and on different OS the problem was not noticed. If you have a problem this time, please specify the browser through which you check it.

**Comment by HenadziStantchik** [comment]
> The problem is the symbol `'`. After checking in different browsers and on different OS the problem was not noticed. If you have a problem this time, please specify the browser through which you check it.

The issue disappeared

**Comment by protosevichVlad** [comment]
Iteration 5. Colorize should work in pdf and html?

File: final_task/rss_reader/rss_reader.py [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+            data = work_with_file.read_feed_form_file(args.date)
+        elif args.source:
+            data = work_with_feedparser.get_object_feed(args.source)
+            data = work_with_dict.to_dict(data)
+            work_with_file.add_feed_to_file(data)
+        else:
+            raise RssReaderException.RssReaderException('How work with application?\nEnter in command line: rss-reader -h')
+
+        if args.version:
+            result = f'RSS reader version {__init__.VERSION}'
```
Comment (dzhigailo) [inline final_task/rss_reader/rss_reader.py]: It's not a good practive to use __init__ file like that. You can define it here, in this module or make a file VERSION.txt, for example, that will contain the version of your application

File: final_task/rss_reader/work_with_feedparser.py [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+import feedparser
+import os
+
+
+@decorators.functions_log
+def get_object_feed(url: str) -> feedparser.FeedParserDict:
+    try:
+        data = feedparser.parse(url)
+        try:
+            if data.status == 200:
```
Comment (HenadziStantchik) [inline final_task/rss_reader/work_with_feedparser.py:12]: It is better to avoid using `status` for checking if we received feed or not.

`data` here has type `FeedParserDict`, which is inherited from `dict`, so it is a dictionary, which has keys and values.
Neither `FeedParserDict` nor `dict` have attribute `status`, so theoretically you should always get an `AttributeError` here.
BUT `FeedParserDict` class has its magic method `__getattribute__` redefined (note: getattribute is called every time you try to access some object attribute using `.`)  in a way so if there is no such attribute (for example `status`), it will try to look in this dictionary keys, using `__getitem__` (note: getitem is called when you try to acccess some item in collection by key or by index: `data["status"]`). So there is no attribute `status`, but `status` is a key in `data` dictionary, so it is still luckily works.

In short, using `status` here is pretty volatile and may break your program, as you can see in one of your tests.
Moreover if you use `bozo` here you will be able to further simplify this function

File: final_task/rss_reader/rss_reader.py [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+    parser.add_argument("--colorize", help="Colorize text", action="store_true")
+    args = parser.parse_args()
+    if not args.limit:
+        args.limit = -1
+    if args.verbose:
+        logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%I:%M:%S %p',
+                            stream=sys.stdout, level=logging.DEBUG)
+    else:
+        logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%I:%M:%S %p',
+                            filename="sample.log", level=logging.DEBUG)
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py:43]: These two logging line are almost the same, the only difference is in 1 keyword argument. It is better to use separate functions in these situations, but it is not mandatory here.

File: final_task/rss_reader/rss_reader.py [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+            data = work_with_file.read_feed_form_file(args.date)
+        elif args.source:
+            data = work_with_feedparser.get_object_feed(args.source)
+            data = work_with_dict.to_dict(data)
+            work_with_file.add_feed_to_file(data)
+        else:
+            raise RssReaderException.RssReaderException('How work with application?\nEnter in command line: rss-reader -h')
+
+        if args.version:
+            result = f'RSS reader version {open("VERSION.txt").readline()}'
```
Comment (HenadziStantchik) [inline final_task/rss_reader/rss_reader.py:67]: It would be better if you put it instead of `pass` before

File: final_task/MANIFEST.in [hunk 28]
```diff
@@ -0,0 +1,2 @@
+include rss_reader\TimesNewRoman.ttf
+include rss_reader\VERSION.txt
```
Comment (HenadziStantchik) [inline final_task/MANIFEST.in:2]: Because Linux and Windows have different filesystem separators, your setup installation will not work on Linux.

**Comment by HenadziStantchik** [comment]
Here is feedback on your app:
Iteration 1:
Everything works fine.
Iteration 2:
Everything seems working (but Windows-only, check my comment on your `MANIFEST` file)
Iteration 3:
Everything works fine, good job on implementing search by rss source.
Iteration 4:
Everything works fine
Iteration 5:
After using `--colorize` once colour of text in command line changes permanently (Linux) it would be nice to have it changed back. Aside from that everything is fine.

One test does not work: `test_rssfile_to_dict`, check my comment on your `work_with_feedparser.py` file. But despite one test not working, coverage rate is still very high.

The code is clean, easy to read, well structured and mostly corresponds to the PEP8 style recommendations. Documentation is a bit lacking, it would be nice if you could put smaller examples here and focus more on describing your project, for example how to install it and etc.

Commit messages mostly correspond to the recommended commit guidelines, but lacking in information.
